### PR TITLE
Add Marlin with act_order support and all group sizes with the ability to load GPTQ models directly

### PR DESCRIPTION
**Pull request by alexm-nm**
1. Adds gptq_marlin that supports act_order == True and all group_sizes
2. Adds on-the-fly marlin_repack from GPTQ weights to Marlin weights
3. 
Tests: Added test_gptq_marlin.py file with tests that cover many configs for 1-gpu and 2-gpu runs

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -202,16 +202,22 @@ def _verify_quantization(self) -> None:
...
                 hf_quant_method = "marlin"
+
+            # Use gptq_marlin if was specified
+            if hf_quant_method == "gptq" and self.quantization == "gptq_marlin":
+                hf_quant_method = "gptq_marlin"
+
             if self.quantization is None:
                 self.quantization = hf_quant_method
-            elif self.quantization != hf_quant_method:
+            elif self.quantization != hf_quant_method:
```
Comment (LucasWilkinson) [inline vllm/config.py:220]: nit: remove formatting change

File: vllm/model_executor/layers/quantization/marlin.py [hunk 2]
```diff
@@ -23,9 +22,8 @@ def __init__(
         self.group_size = group_size
         if self.group_size != 128 and self.group_size != -1:
             raise ValueError(
-                "Currently, only group size 128 and -1 (channelwise) "
-                "is supported for Marlin, but got group_size of "
-                f"{self.group_size}")
+                "Currently, only group size 128 and -1 (channelwise) is supported for "
+                f"Marlin, but got group_size of {self.group_size}")
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/marlin.py:26]: nit: why the format change?

File: vllm/model_executor/layers/quantization/gptq_marlin.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        scales_and_zp_size = input_size // group_size
+        scales_and_zp_input_dim = None
+
+        if self.quant_config.desc_act:
+            assert self.quant_config.group_size != -1
+
+            is_k_full = input_size_per_partition == input_size
+
+        else:
+            is_k_full = True  # Because of full alignment with group-size and shard of scales/zp
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/quantization/gptq_marlin.py:160]: nit: should we just drop the `else` since `is_k_full` is already `True` by default?

**Review by LucasWilkinson — APPROVED** [review]
LGTM (once green), the only thing I would maybe consider is if its possible to merge `MarlinConfig` and `GPTQMarlinConfig`, is there any reason we can't make `GPTQMarlinConfig` just an instance of `MarlinConfig` with a  `repack_from_gptq` flag set which could be set in `MarlinConfig.from_config` by detecting that `is_marlin_format` is not set, then we could merge `GPTQMarlinLinearMethod` and `MarlinLinearMethod` too (and the user would just set the quantization to `marlin` but pass a `gptq` model, i.e.:
```
 # Use gptq_marlin if was specified
             if hf_quant_method == "gptq" and self.quantization == "gptq_marlin":
                 hf_quant_method = "gptq_marlin"
```
becomes:
```
 # Use gptq_marlin if was specified
             if hf_quant_method == "gptq" and self.quantization == "marlin":
                 hf_quant_method = "marlin"
```
)

not sure if this would cause a slew of other issues

**Review by robertgshaw2-neuralmagic — CHANGES_REQUESTED** [review]
The key here is that the user should not need to know about Marlin. They should just pass a GPTQ model and get performance from Marlin

So the UX should be:

#### If user passes nothing to the `--quantization` argument

A) If `hf_config.quant_method == gptq` AND `can_be_converted_to_marlin` AND `gpu_supports_marlin`, we should repack and use the marlin kernels

B) If `hf_config.quant_method == qptq` and `hf_config.is_marlin_format=True`, we should just use the marlin kernels (if the gpu supports, else fail)

C) else, use the exllama kernels


#### If the user passes something to the `--quantization` argument, we should follow what they say to do

File: vllm/config.py (hunk 1 above)
Comment (alexm-nm) [reply vllm/config.py:220]: done

File: vllm/model_executor/layers/quantization/marlin.py (hunk 2 above)
Comment (alexm-nm) [reply vllm/model_executor/layers/quantization/marlin.py]: due to format.sh

File: vllm/model_executor/layers/quantization/gptq_marlin.py (hunk 3 above)
Comment (alexm-nm) [reply vllm/model_executor/layers/quantization/gptq_marlin.py]: I removed the init to True, since each part has its own logic.

**Comment by alexm-nm** [comment]
Merging marlin config + linear with gptq marlin config + linear is complicated because there are many differences (they invoke different GPU kernels that have a different set of restrictions and features). We can do it later after we decide to remove the marlin models completely, since we could only let users go via the gptq path and apply marlin if it can be used (as Rob said)

I will modify the interface of --quantization to what @robertgshaw2-neuralmagic describes, I think it is a good idea to apply marlin if it can be applied.

**Comment by alexm-nm** [comment]
Modified the interface to auto apply gptq_marlin when it is possible.

**Comment by robertgshaw2-neuralmagic** [comment]
Can you please add some testing to this PR?

Specifically, would like to see E2E validation that we can take a GPTQ model, convert and load it in Marlin, and get correct inference results. This would be nice to see for various group sizes and various act_order

Additionally, would like to see some multi-gpu tests that show correctness in the same scenario

Thanks!

**Comment by alexm-nm** [comment]
Good idea, adding E2E tests.

**Comment by robertgshaw2-neuralmagic** [comment]
Closing in favor of upstream

**Pull request closed without merging**